File: internal/outreach/run_summary.go
package outreach

import (
	"fmt"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

type RunSummary struct {
	RunID            string             `json:"run_id"`
	Command          string             `json:"command"`
	Args             []string           `json:"args,omitempty"`
	StartedAt        time.Time          `json:"started_at"`
	CompletedAt      time.Time          `json:"completed_at,omitempty"`
	Status           string             `json:"status"`
	Blocker          string             `json:"blocker,omitempty"`
	DashboardPath    string             `json:"dashboard_path,omitempty"`
	StatePath        string             `json:"state_path,omitempty"`
	TargetAgencies   int                `json:"target_agencies"`
	TargetRecruiters int                `json:"target_recruiters"`
	AllowSend        bool               `json:"allow_send"`
	Counts           DashboardRunCounts `json:"counts"`
	Actions          []DailyLeadAction  `json:"actions,omitempty"`
	Recommendation   RunRecommendation  `json:"recommendation"`
}

type RunRecommendation struct {
	ShouldRetry bool   `json:"should_retry"`
	Command     string `json:"command,omitempty"`
	Reason      string `json:"reason"`
	Blocker     string `json:"blocker,omitempty"`
}

func normalizeRunID(value string, prefix string) string {
	cleaned := cleanText(value)
	if cleaned != "" {
		return cleaned
	}
	return newRunID(prefix)
}

func newRunID(prefix string) string {
	cleanPrefix := strings.ToLower(cleanText(prefix))
	if cleanPrefix == "" {
		cleanPrefix = "run"
	}
	cleanPrefix = strings.NewReplacer(" ", "-", "_", "-").Replace(cleanPrefix)
	return fmt.Sprintf("%s-%s", cleanPrefix, time.Now().Format("20060102T150405.000000000"))
}

func (s Store) RunDashboardPath(runID string) string {
	return filepath.Join(s.Dir, "dashboards", "runs", normalizeRunID(runID, "run")+".md")
}

func (s Store) LatestRunDashboardPath() string {
	return filepath.Join(s.Dir, "dashboards", "latest-run.md")
}

func (s Store) LatestRenderDashboardPath() string {
	return filepath.Join(s.Dir, "dashboards", "latest-render.md")
}

func appendRunLifecycleEvent(store *Store, event RunEvent) error {
	state, err := store.Load()
	if err != nil {
		return err
	}
	appendRunEvent(&state, event)
	return store.Save(state)
}

func LatestRunSummary(state OutreachState, statePath string) (RunSummary, bool) {
	state.Normalize()
	runs := runSummariesFromEvents(state.RunEvents, statePath)
	if len(runs) == 0 {
		return legacyRunSummaryFromEvents(state.RunEvents, statePath)
	}
	sort.SliceStable(runs, func(i, j int) bool {
		return effectiveRunTime(runs[i]).After(effectiveRunTime(runs[j]))
	})
	return runs[0], true
}

func legacyRunSummaryFromEvents(events []RunEvent, statePath string) (RunSummary, bool) {
	actions := []DailyLeadAction{}
	var startedAt time.Time
	var completedAt time.Time
	for _, event := range events {
		if event.Phase != "send-message" {
			continue
		}
		if startedAt.IsZero() || event.At.Before(startedAt) {
			startedAt = event.At
		}
		if completedAt.IsZero() || event.At.After(completedAt) {
			completedAt = event.At
		}
		actions = append(actions, runEventAction(event))
	}
	if len(actions) == 0 {
		return RunSummary{}, false
	}
	summary := RunSummary{
		RunID:            "legacy-" + completedAt.Format("20060102T150405"),
		Command:          "legacy-run-events",
		StartedAt:        startedAt,
		CompletedAt:      completedAt,
		Status:           "completed",
		StatePath:        statePath,
		TargetAgencies:   5,
		TargetRecruiters: 5,
		AllowSend:        true,
		Actions:          actions,
		Counts:           dashboardRunCounts(actions),
	}
	summary.Recommendation = RecommendNextRunSummary(summary)
	return summary, true
}

func runSummariesFromEvents(events []RunEvent, statePath string) []RunSummary {
	byID := map[string]*RunSummary{}
	for _, event := range events {
		if cleanText(event.RunID) == "" {
			continue
		}
		summary := byID[event.RunID]
		if summary == nil {
			summary = &RunSummary{
				RunID:     event.RunID,
				Status:    "running",
				StatePath: statePath,
			}
			byID[event.RunID] = summary
		}
		if !event.StartedAt.IsZero() {
			summary.StartedAt = event.StartedAt
		}
		if summary.StartedAt.IsZero() && event.Phase == "run-start" {
			summary.StartedAt = event.At
		}
		if !event.CompletedAt.IsZero() {
			summary.CompletedAt = event.CompletedAt
		}
		if event.Command != "" {
			summary.Command = event.Command
		}
		if len(event.Args) > 0 {
			summary.Args = append([]string(nil), event.Args...)
		}
		if event.DashboardPath != "" {
			summary.DashboardPath = event.DashboardPath
		}
		if event.StatePath != "" {
			summary.StatePath = event.StatePath
		}
		if event.TargetAgencies != 0 {
			summary.TargetAgencies = event.TargetAgencies
		}
		if event.TargetRecruiters != 0 {
			summary.TargetRecruiters = event.TargetRecruiters
		}
		if event.AllowSend {
			summary.AllowSend = true
		}
		if event.Blocker != "" {
			summary.Blocker = event.Blocker
		}
		if event.Phase == "send-message" {
			action := runEventAction(event)
			summary.Actions = append(summary.Actions, action)
		}
		if event.Phase == "run-finish" {
			if event.Result != "" {
				summary.Status = event.Result
			} else {
				summary.Status = "completed"
			}
		}
	}
	runs := make([]RunSummary, 0, len(byID))
	for _, summary := range byID {
		summary.Counts = dashboardRunCounts(summary.Actions)
		summary.Recommendation = RecommendNextRunSummary(*summary)
		runs = append(runs, *summary)
	}
	return runs
}

func effectiveRunTime(summary RunSummary) time.Time {
	if !summary.CompletedAt.IsZero() {
		return summary.CompletedAt
	}
	return summary.StartedAt
}

func runEventAction(event RunEvent) DailyLeadAction {
	return DailyLeadAction{
		At:         event.At,
		RunID:      event.RunID,
		Bucket:     event.Bucket,
		LeadID:     event.LeadID,
		Name:       event.Name,
		Action:     "send-message",
		Result:     event.Result,
		Note:       optionalString(event.Note),
		ProfileURL: nil,
	}
}

func optionalString(value string) *string {
	if cleanText(value) == "" {
		return nil
	}
	return &value
}

func RecommendNextRun(state OutreachState, statePath string, targetAgencies int, targetRecruiters int, allowSend bool) RunRecommendation {
	if summary, ok := LatestRunSummary(state, statePath); ok {
		if summary.TargetAgencies == 0 {
			summary.TargetAgencies = targetAgencies
		}
		if summary.TargetRecruiters == 0 {
			summary.TargetRecruiters = targetRecruiters
		}
		if allowSend {
			summary.AllowSend = true
		}
		return RecommendNextRunSummary(summary)
	}
	if targetAgencies > 0 {
		return RunRecommendation{
			ShouldRetry: true,
			Command:     retryCommand(targetAgencies, 0, allowSend),
			Reason:      "No previous run summary is available. Start with an agency-focused run if agency coverage is the open question.",
		}
	}
	return RunRecommendation{Reason: "No previous run summary is available."}
}

func RecommendNextRunSummary(summary RunSummary) RunRecommendation {
	agencyGap := nonZero(summary.TargetAgencies, 0) - summary.Counts.Sent.Agencies
	recruiterGap := nonZero(summary.TargetRecruiters, 0) - summary.Counts.Sent.Recruiters
	if summary.Status == "failed" || cleanText(summary.Blocker) != "" {
		targetAgencies := positiveOrFallback(agencyGap, nonZero(summary.TargetAgencies, 5))
		targetRecruiters := positiveOrFallback(recruiterGap, nonZero(summary.TargetRecruiters, 5))
		if agencyGap > 0 && recruiterGap <= 0 {
			targetRecruiters = 0
		} else if recruiterGap > 0 && agencyGap <= 0 {
			targetAgencies = 0
		}
		return RunRecommendation{
			ShouldRetry: true,
			Command:     retryCommand(targetAgencies, targetRecruiters, summary.AllowSend),
			Reason:      "The latest run did not finish cleanly.",
			Blocker:     summary.Blocker,
		}
	}
	if summary.AllowSend {
		if agencyGap > 0 {
			return RunRecommendation{
				ShouldRetry: true,
				Command:     retryCommand(agencyGap, 0, true),
				Reason:      fmt.Sprintf("Agency target is still short by %d sends; validate the fixed agency lane without spending time on recruiters.", agencyGap),
			}
		}
		if recruiterGap > 0 {
			return RunRecommendation{
				ShouldRetry: true,
				Command:     retryCommand(0, recruiterGap, true),
				Reason:      fmt.Sprintf("Recruiter target is still short by %d sends.", recruiterGap),
			}
		}
	}
	return RunRecommendation{Reason: "Latest run reached its requested send target; no retry is needed for the same target."}
}

func retryCommand(targetAgencies int, targetRecruiters int, allowSend bool) string {
	parts := []string{
		"/Users/hanifcarroll/.local/bin/recruiter-agency-outreach",
		"run-daily",
		"--session", "auto",
		"--target-agencies", fmt.Sprintf("%d", targetAgencies),
		"--target-recruiters", fmt.Sprintf("%d", targetRecruiters),
		"--refresh-saved-searches",
		"--print-markdown",
	}
	if allowSend {
		parts = append(parts, "--allow-send")
	}
	if targetAgencies > 0 {
		parts = append(parts, "--timeout-ms", "240000", "--stop-when-no-progress", "--max-no-progress-searches", "12")
	}
	return strings.Join(parts, " ")
}

func nonZero(value int, fallback int) int {
	if value != 0 {
		return value
	}
	return fallback
}

func positiveOrFallback(value int, fallback int) int {
	if value > 0 {
		return value
	}
	return fallback
}

func RenderRunSummaryText(summary RunSummary) string {
	lines := []string{
		fmt.Sprintf("run_id=%s", summary.RunID),
		fmt.Sprintf("status=%s", summary.Status),
		fmt.Sprintf("state=%s", summary.StatePath),
		fmt.Sprintf("dashboard=%s", summary.DashboardPath),
		fmt.Sprintf("target=%d agencies,%d recruiters", summary.TargetAgencies, summary.TargetRecruiters),
		fmt.Sprintf("sent=%d agencies,%d recruiters", summary.Counts.Sent.Agencies, summary.Counts.Sent.Recruiters),
		fmt.Sprintf("checked_skipped=conversation_exists %d agencies,%d recruiters; not_messageable %d agencies,%d recruiters; blocked %d agencies,%d recruiters; send_failed %d agencies,%d recruiters",
			summary.Counts.ConversationExists.Agencies,
			summary.Counts.ConversationExists.Recruiters,
			summary.Counts.NotMessageable.Agencies,
			summary.Counts.NotMessageable.Recruiters,
			summary.Counts.Blocked.Agencies,
			summary.Counts.Blocked.Recruiters,
			summary.Counts.SendFailed.Agencies,
			summary.Counts.SendFailed.Recruiters,
		),
	}
	if cleanText(summary.Blocker) != "" {
		lines = append(lines, "blocker="+cleanText(summary.Blocker))
	}
	if summary.Recommendation.ShouldRetry {
		lines = append(lines, "recommendation="+summary.Recommendation.Reason)
		lines = append(lines, "next_command="+summary.Recommendation.Command)
	} else if cleanText(summary.Recommendation.Reason) != "" {
		lines = append(lines, "recommendation="+summary.Recommendation.Reason)
	}
	return strings.Join(lines, "\n")
}

File: apps/recruiter_agency_outreach/run_summary.py
"""Run summary and retry recommendation helpers for recruiter/agency outreach."""

from __future__ import annotations

from dataclasses import dataclass, field

from .dashboard import (
    BucketCounts,
    DailyLeadAction,
    RunCounts,
    bucket_for_lead,
    dashboard_run_counts,
)
from .models import Lead, LeadType, MessageStatus, OutreachState, RunEvent
from .sourcing import find_lead_by_id
from .utils import clean_text, now_iso


@dataclass(slots=True)
class RunRecommendation:
    should_retry: bool = False
    command: str = ""
    reason: str = ""
    blocker: str = ""


@dataclass(slots=True)
class RunSummary:
    run_id: str
    command: str = ""
    args: list[str] = field(default_factory=list)
    started_at: str = ""
    completed_at: str = ""
    status: str = "running"
    blocker: str = ""
    dashboard_path: str = ""
    state_path: str = ""
    target_agencies: int = 0
    target_recruiters: int = 0
    allow_send: bool = False
    counts: RunCounts = field(default_factory=RunCounts)
    actions: list[DailyLeadAction] = field(default_factory=list)
    recommendation: RunRecommendation = field(default_factory=RunRecommendation)


def new_run_id(prefix: str) -> str:
    cleaned_prefix = clean_text(prefix).lower().replace(" ", "-").replace("_", "-") or "run"
    stamp = now_iso().replace(":", "").replace("-", "").replace(".", "")
    return f"{cleaned_prefix}-{stamp}"


def latest_run_summary(state: OutreachState, state_path: str) -> RunSummary | None:
    summaries = _run_summaries_from_events(state, state_path)
    if not summaries:
        return None
    summaries.sort(key=_effective_run_time, reverse=True)
    return summaries[0]


def recommend_next_run(
    state: OutreachState,
    state_path: str,
    *,
    target_agencies: int,
    target_recruiters: int,
    allow_send: bool,
) -> RunRecommendation:
    summary = latest_run_summary(state, state_path)
    if summary is not None:
        if summary.target_agencies == 0:
            summary.target_agencies = target_agencies
        if summary.target_recruiters == 0:
            summary.target_recruiters = target_recruiters
        if allow_send:
            summary.allow_send = True
        return recommend_next_run_summary(summary)
    if target_agencies > 0:
        return RunRecommendation(
            should_retry=True,
            command=_retry_command(target_agencies, 0, allow_send),
            reason=(
                "No previous run summary is available. Start with an agency-focused "
                "run if agency coverage is the open question."
            ),
        )
    return RunRecommendation(reason="No previous run summary is available.")


def recommend_next_run_summary(summary: RunSummary) -> RunRecommendation:
    agency_gap = _non_zero(summary.target_agencies, 0) - summary.counts.sent.agencies
    recruiter_gap = _non_zero(summary.target_recruiters, 0) - summary.counts.sent.recruiters
    if summary.status == "failed" or clean_text(summary.blocker):
        target_agencies = _positive_or_default(agency_gap, _non_zero(summary.target_agencies, 5))
        target_recruiters = _positive_or_default(
            recruiter_gap,
            _non_zero(summary.target_recruiters, 5),
        )
        if agency_gap > 0 and recruiter_gap <= 0:
            target_recruiters = 0
        elif recruiter_gap > 0 and agency_gap <= 0:
            target_agencies = 0
        return RunRecommendation(
            should_retry=True,
            command=_retry_command(target_agencies, target_recruiters, summary.allow_send),
            reason="The latest run did not finish cleanly.",
            blocker=summary.blocker,
        )
    if summary.allow_send:
        if agency_gap > 0:
            return RunRecommendation(
                should_retry=True,
                command=_retry_command(agency_gap, 0, True),
                reason=(
                    f"Agency target is still short by {agency_gap} sends; validate "
                    "the fixed agency lane without spending time on recruiters."
                ),
            )
        if recruiter_gap > 0:
            return RunRecommendation(
                should_retry=True,
                command=_retry_command(0, recruiter_gap, True),
                reason=f"Recruiter target is still short by {recruiter_gap} sends.",
            )
    if not summary.allow_send:
        return RunRecommendation(
            reason=(
                "Latest sourcing run completed. Use send-ready for already "
                "dry_run_ready leads, or continue sourcing if the dashboard shows "
                "a readiness shortfall."
            )
        )
    return RunRecommendation(
        reason="Latest sending run reached its requested send target; no retry is needed."
    )


def render_run_summary_text(summary: RunSummary) -> str:
    lines = [
        f"run_id={summary.run_id}",
        f"status={summary.status}",
        f"state={summary.state_path}",
        f"dashboard={summary.dashboard_path}",
        f"target={summary.target_agencies} agencies,{summary.target_recruiters} recruiters",
        f"sent={summary.counts.sent.agencies} agencies,{summary.counts.sent.recruiters} recruiters",
        (
            "checked_skipped="
            f"conversation_exists {summary.counts.conversation_exists.agencies} agencies,"
            f"{summary.counts.conversation_exists.recruiters} recruiters; "
            f"not_messageable {summary.counts.not_messageable.agencies} agencies,"
            f"{summary.counts.not_messageable.recruiters} recruiters; "
            f"blocked {summary.counts.blocked.agencies} agencies,"
            f"{summary.counts.blocked.recruiters} recruiters; "
            f"send_failed {summary.counts.send_failed.agencies} agencies,"
            f"{summary.counts.send_failed.recruiters} recruiters"
        ),
    ]
    if clean_text(summary.blocker):
        lines.append(f"blocker={clean_text(summary.blocker)}")
    if summary.recommendation.should_retry:
        lines.append(f"recommendation={summary.recommendation.reason}")
        lines.append(f"next_command={summary.recommendation.command}")
    elif clean_text(summary.recommendation.reason):
        lines.append(f"recommendation={summary.recommendation.reason}")
    return "\n".join(lines)


def daily_action_from_lead(
    lead: Lead,
    *,
    run_id: str,
    bucket: str,
    action: str,
    result: str,
    note: str | None = None,
    at: str | None = None,
) -> DailyLeadAction:
    return DailyLeadAction(
        at=at or now_iso(),
        run_id=run_id,
        bucket=bucket,
        lead_id=lead.id,
        name=lead.name,
        profile_url=lead.profile_url,
        lead_type=lead.lead_type,
        message_status=lead.message_status,
        action=action,
        result=result,
        note=note,
    )


def sent_count_from_actions(actions: list[DailyLeadAction], bucket: str) -> int:
    return sum(
        1 for action in actions if action.bucket == bucket and action.result == "sent-clicked"
    )


def _run_summaries_from_events(
    state: OutreachState,
    state_path: str,
) -> list[RunSummary]:
    by_run_id: dict[str, RunSummary] = {}
    for event in state.run_events:
        if not clean_text(event.run_id):
            continue
        summary = by_run_id.setdefault(
            event.run_id,
            RunSummary(run_id=event.run_id, state_path=state_path),
        )
        if event.started_at:
            summary.started_at = event.started_at
        if not summary.started_at and event.phase == "run-start":
            summary.started_at = event.at
        if event.completed_at:
            summary.completed_at = event.completed_at
        if event.command:
            summary.command = event.command
        if event.args:
            summary.args = list(event.args)
        if event.dashboard_path:
            summary.dashboard_path = event.dashboard_path
        if event.state_path:
            summary.state_path = event.state_path
        if event.target_agencies:
            summary.target_agencies = event.target_agencies
        if event.target_recruiters:
            summary.target_recruiters = event.target_recruiters
        if event.allow_send:
            summary.allow_send = True
        if event.blocker:
            summary.blocker = event.blocker
        if event.phase == "send-message":
            summary.actions.append(_run_event_action(state, event))
        if event.phase == "run-finish":
            summary.status = event.result or "completed"
    summaries = list(by_run_id.values())
    for summary in summaries:
        summary.counts = dashboard_run_counts(summary.actions)
        summary.recommendation = recommend_next_run_summary(summary)
    return summaries


def _run_event_action(state: OutreachState, event: RunEvent) -> DailyLeadAction:
    lead = find_lead_by_id(state.leads, event.lead_id)
    if lead is None:
        return DailyLeadAction(
            at=event.at,
            run_id=event.run_id,
            bucket=event.bucket,
            lead_id=event.lead_id,
            name=event.name,
            profile_url=None,
            lead_type=LeadType.BAD_FIT,
            message_status=MessageStatus.NONE,
            action="send-message",
            result=event.result,
            note=event.note or None,
        )
    return daily_action_from_lead(
        lead,
        run_id=event.run_id,
        bucket=event.bucket or bucket_for_lead(lead),
        action="send-message",
        result=event.result or lead.message_status.value,
        note=event.note or None,
        at=event.at,
    )


def _effective_run_time(summary: RunSummary) -> str:
    return summary.completed_at or summary.started_at


def _retry_command(target_agencies: int, target_recruiters: int, allow_send: bool) -> str:
    if allow_send:
        return (
            "recruiter-agency-outreach send-ready --session auto "
            f"--target-agencies {target_agencies} "
            f"--target-recruiters {target_recruiters} "
            "--allow-send --print-markdown"
        )
    return (
        "recruiter-agency-outreach run-daily --session auto "
        f"--target-agencies {target_agencies} "
        f"--target-recruiters {target_recruiters} "
        "--print-markdown"
    )


def _non_zero(value: int, default: int) -> int:
    return value if value != 0 else default


def _positive_or_default(value: int, default: int) -> int:
    return value if value > 0 else default


def empty_run_counts() -> RunCounts:
    return RunCounts(
        sent=BucketCounts(),
        dry_run_ready=BucketCounts(),
        conversation_exists=BucketCounts(),
        not_messageable=BucketCounts(),
        blocked=BucketCounts(),
        send_failed=BucketCounts(),
    )
